File: daily challenge/2024/January/872. Leaf-Similar Trees.go
package main

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func leafSimilar(root1 *TreeNode, root2 *TreeNode) bool {
	list1 := dfs(root1)
	list2 := dfs(root2)
	if len(list1) != len(list2) {
		return false
	}

	for i := 0; i < len(list1); i++ {
		if list1[i] != list2[i] {
			return false
		}
	}
	return true
}

func dfs(node *TreeNode) []int {
	stack := []*TreeNode{}
	list := []int{}
	stack = append(stack, node)
	for len(stack) != 0 {
		top := stack[len(stack)-1]
		stack = stack[:len(stack)-1]
		if top.Left == nil && top.Right == nil {
			list = append(list, top.Val)
		}
		if top.Left != nil {
			stack = append(stack, top.Left)
		}
		if top.Right != nil {
			stack = append(stack, top.Right)
		}
	}

	return list
}

File: 872.leaf-similar-trees.ts
/*
 * @lc app=leetcode id=872 lang=typescript
 *
 * [872] Leaf-Similar Trees
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */

function leafSimilar(root1: TreeNode | null, root2: TreeNode | null): boolean {
  const leaf1: number[] = [];
  const leaf2: number[] = [];

  function dfs(root: TreeNode | null, pool: number[]) {
    if (root === null) {
      return;
    }
    if (!root.left && !root.right) {
      pool.push(root.val);
    }

    dfs(root.left, pool);
    dfs(root.right, pool);
  }

  dfs(root1, leaf1);
  dfs(root2, leaf2);
  if (leaf1.length !== leaf2.length) {
    return false;
  }

  for (let i = 0; i < leaf1.length; i++) {
    if (leaf1[i] !== leaf2[i]) {
      return false;
    }
  }

  return true;
}
// @lc code=end
